### PR TITLE
Language Models/Benchmarks Integration 

### PR DESCRIPTION
**Pull request by mike-ferguson**
This PR will (hopefully) allow Brain-Score to now support language models as well as vision. Still a lot todo: 

- [x] Make fixtures more accurate (actual extra fields needed in meta for lang models)
- [x] Only show lang models/benchmarks for lang models in leaderboard, same for vision
- [x] Submission system frontend (one submission page with checkbox of lang/vision model, or two submission pages?)
- [x] Potentially rename parent benchmarks to vision_average, vision_engineering, to create language_average, language_engineering, etc. 
- [x] Create new JSON data when submitting model (with lang vs vision flag, will be similar to the competition change). This is what Jenkins will use to run the appropriate env setup/scoring script. 
- [x] Determine what (if there are) neural language sites, and types of lang benchmarks (_what goes in leaderboard columns_)
- [x] Edit URLs to point to correct page to render (not just index.html)
- [x] Add dev support for new Openmind environments @mschrimpf what you did to the run_benchmarks job needs to be done with the dev_run_benchmarks job

Beta test TODO:

- [x] Collapse engineering header
- [x] Update benchmark type names
- [ ] Bibtex for language

### CODE REVIEW

**Comment by mschrimpf** [comment]
re 
> Add dev support for new Openmind environments @mschrimpf what you did to the run_benchmarks job needs to be done with the dev_run_benchmarks job

Not sure what you mean, I haven't changed `run_benchmarks`

File: benchmarks/models.py [hunk 1]
```diff
@@ -176,6 +176,7 @@ class Model(models.Model):
     visual_degrees = models.IntegerField(null=True)  # null during model creation before querying objec
     public = models.BooleanField(default=False)
     competition = models.CharField(max_length=200, default=None, null=True)
+    domain = models.CharField(max_length=200, default="vision", null=True)
```
Comment (mschrimpf) [inline benchmarks/models.py:179]:
```suggestion
    domain = models.CharField(max_length=200)
```
This should always be specified, and cannot be null

File: benchmarks/templates/benchmarks/compare.html [hunk 2]
```diff
@@ -4,14 +4,26 @@
     <div class="xlabel-container">
         <select id="xlabel">
             {% for benchmark in benchmarks %}
-                <option value="{{ benchmark.identifier }}">{{ benchmark.short_name }}</option>
+                {% if "average" in benchmark.short_name %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/compare.html:7]:
```suggestion
                {% if "average_" in benchmark.short_name %}  {# hide the domain suffix, e.g. `average_vision` becomes `average` #}
```

File: benchmarks/templates/benchmarks/compare.html [hunk 3]
```diff
@@ -4,14 +4,26 @@
...
         <select id="ylabel">
             {% for benchmark in benchmarks %}
-                <option value="{{ benchmark.identifier }}">{{ benchmark.short_name }}</option>
+                {% if "average" in benchmark.short_name %}
+                    <option value="{{ benchmark.identifier }}">average</option>
+                {% elif "engineering" in benchmark.short_name %}
+                    <option value="{{ benchmark.identifier }}">engineering</option>
+                {% else %}
+                    <option value="{{ benchmark.identifier }}">{{ benchmark.short_name }}</option>
+                {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/compare.html:26]: This (and the section above) would be more cleanly handled in a template filter, e.g. 
`<option value="{{ benchmark.identifier }}">{{ benchmark.short_name|simplify_domain }}</option>`
where
```python
def simplify_domain(benchmark_name:str) -> str:
    suffixed_benchmarks = ['average', 'engineering']
    for suffixed_name in suffixed_benchmarks:
        if benchmark_name.startswith(f"{suffixed_name}_"):
            return suffixed_name
    return benchmark_name
```

File: benchmarks/templates/benchmarks/domain-information.html [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                            <p class="control">
+                                <input class="input" type="text" name="display_name" value="{{ user.display_name }}">
+                            </p>
+
+                            <p class="control">
+                                <input type="submit" value="Save" class="button">
+                            </p>
+                        </div>
+                    </form>
+                {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/domain-information.html:37]: redundant and not scalable. Do away with these `if-else` distinctions and make it variable dependent.

File: benchmarks/templates/benchmarks/invalid_zip.html [hunk 5]
```diff
@@ -0,0 +1,18 @@
+{% extends 'benchmarks/base.html' %}
+
+{% block content %}
+<section id="success" class="container center">
+  <h2>Uh Oh! Your submission package was not in the correct format.</h2>
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/invalid_zip.html:5]:
```suggestion
  <h2>Your submission package was not in the correct format.</h2>
```
Let's keep some professionalism

File: benchmarks/templates/benchmarks/invalid_zip.html [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  <p>
+    {{ error }}
+  </p>
+  <h3>The submission packaging rules can be found
+    <a href="https://brain-score.readthedocs.io/en/latest/">here</a></h3>
+  {% if "language" in request.path %}
+    <h5>To resubmit, click <a href="http://{{ request.get_host }}/profile/language/submit">here</a>.</h5>
+  {% else %}
+    <h5>To resubmit, click <a href="http://{{ request.get_host }}/profile/vision/submit">here</a>.</h5>
+  {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/invalid_zip.html:15]: use variable instead of if-else

File: benchmarks/templates/benchmarks/login.html [hunk 7]
```diff
@@ -27,13 +28,26 @@ <h1>Login</h1>
...
             {% endfor %}
-            <p style="font-size: 12px"><a href="/password">Forgot Password</a></p>
-            <button type="button submit" class="button submission">Login</button>
+            {% if "language" in request.path %}
+                <p style="font-size: 12px"><a href="/password/language">Forgot Password</a></p>
+                <button type="button submit" class="button submission">Login</button>
+            {% else %}
+                <p style="font-size: 12px"><a href="/password/vision">Forgot Password</a></p>
+                <button type="button submit" class="button submission">Login</button>
+            {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/login.html:37]: use variable instead of if-else

File: benchmarks/templates/benchmarks/login.html [hunk 8]
```diff
@@ -27,13 +28,26 @@ <h1>Login</h1>
...
             <p style="color: red;">The username or password was incorrect</p>
         {% endif %}
         <br/>
-        <button class="button"><a href='/signup'>Create account</a></button>
+
+        {% if "language" in request.path %}
+            <button class="button"><a href='/signup/language'>Create account</a></button>
+        {% else %}
+            <button class="button"><a href='/signup/vision'>Create account</a></button>
+        {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/login.html:50]: use variable instead of if-else

File: benchmarks/templates/benchmarks/participate.html [hunk 9]
```diff
@@ -22,35 +23,40 @@ <h4 class="title is-4">Change the evaluation: Submit a metric</h4>
             </div>
 
             <h4 class="title is-4">Citation</h4>
-            <div class="block">
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/participate.html:25]: undo this change, keep what was before for now

File: benchmarks/templates/benchmarks/password-confirm.html [hunk 10]
```diff
@@ -3,7 +3,12 @@
 {% block content %}
 <section id="password" class="container center">
   <p>An email was sent. </p>
-  <p><a href="../profile/">Return to login</a>.</p>
+
+  {% if "language" in request.path %}
+    <h5><a href="/profile/language">Return to Login</a>.</h5>
+  {% else %}
+    <h5><a href="/profile/vision">Return to Login</a>.</h5>
+  {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/password-confirm.html:11]: use variable instead of if-else

File: benchmarks/templates/benchmarks/password.html [hunk 11]
```diff
@@ -21,6 +21,10 @@ <h2>Change Password</h2>
         </form>
 
         <br/>
-        <p><a href="/../profile/">Return to login</a></p>
+        {% if "language" in request.path %}
+            <h5><a href="/profile/language">Return to Login</a>.</h5>
+        {% else %}
+            <h5><a href="/profile/vision">Return to Login</a>.</h5>
+        {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/password.html:28]: use variable instead of if-else

File: benchmarks/templates/benchmarks/profile.html [hunk 12]
```diff
@@ -6,38 +6,50 @@
...
+                    <div class="field has-addons">
+                        <p class="control">
+                            <input class="input" type="text" name="display_name" value="{{ user.display_name }}">
+                        </p>
+                        <p class="control">
+                            <input type="submit" value="Save" class="button">
+                        </p>
+                    </div>
+                </form>
+            {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/profile.html:37]: use variable instead of if-else

File: benchmarks/templates/benchmarks/signup.html [hunk 13]
```diff
@@ -20,6 +20,12 @@ <h2>Create new account</h2>
             <button type="submit" class="button submission">Sign up</button>
         </form>
         <br/>
-        <button class="button"><a href="/../profile/">Return to login</a></button>
+
+        {% if "language" in request.path %}
+            <button class="button"><a href="/../profile/language">Return to login</a></button>
+        {% else %}
+            <button class="button"><a href="/../profile/vision">Return to login</a></button>
+        {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/signup.html:28]: use variable instead of if-else

File: benchmarks/templates/benchmarks/submissions.html [hunk 14]
```diff
@@ -2,28 +2,71 @@
...
     <h2>Resubmit models</h2>
     <p>Rerun models on new benchmark version. Benchmarks are only executed when no score available.</p>
-    <form action="{% url 'resubmit' %}" method="post">
-        {% csrf_token %}
-        <div class="columns">
-            <div class="column is-half">
-                <h4>Pick benchmarks</h4>
-                {% for short_name, benchmark_type_id in submittable_benchmarks.items %}
+
+    {% if "language" in request.path %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/submissions.html:6]: use variable instead of if-else

File: benchmarks/templates/benchmarks/table.html [hunk 15]
```diff
@@ -67,13 +67,19 @@
...
-                                   {% if benchmark.root_parent == 'average' %} brain_benchmark {% endif %}">
-                        {{ benchmark.short_name }}
+                                   {% if benchmark.root_parent == 'average_vision' %} brain_benchmark {% endif %}">
+                        {% if "average" in benchmark.short_name %}
+                            average
+                        {% elif "engineering" in benchmark.short_name %}
+                            engineering
+                        {% else %}
+                            {{ benchmark.short_name }}
+                        {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/table.html:77]: use template tag (see comment above)

File: benchmarks/templates/benchmarks/table.html [hunk 16]
```diff
@@ -67,13 +67,19 @@
...
                         </span>
                         {# version #}
                         {% include "benchmarks/benchmark_version.html" with version=benchmark.version %}
                         {# number of child benchmarks, ignore roots #}
-                        {% if benchmark.identifier in uniform_parents and benchmark.short_name != 'average' %}
+                        {% if benchmark.identifier in uniform_parents and benchmark.short_name != 'average_vision' %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/table.html:82]: why does this just apply to vision?

File: benchmarks/templates/benchmarks/table.html [hunk 17]
```diff
@@ -99,9 +105,15 @@
...
-                        </a>
+                         {% if "language" in request.path %}
+                              <a target="_self" rel="noopener noreferrer" href="{% url 'model-language' model.id %}">
+                                {{ model|display_model:user }}
+                            </a>
+                        {% else %}
+                            <a target="_self" rel="noopener noreferrer" href="{% url 'model-vision' model.id %}">
+                                {{ model|display_model:user }}
+                            </a>
+                        {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/table.html:116]: use variable instead of if-else

File: benchmarks/templates/benchmarks/upload.html [hunk 18]
```diff
@@ -3,9 +3,24 @@
 
 {% block content %}
     <section id="upload" class="container center">
-        <h1>Model Submission</h1>
-        You can upload files here to test. For an example of what your submission should include, please visit <a
-            href="https://github.com/brain-score/sample-model-submission/">this template</a>.<br>
+
+        {% if "language" in request.path %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/upload.html:7]: Here it's actually okay for the moment

File: benchmarks/templates/benchmarks/upload.html [hunk 19]
```diff
@@ -27,7 +42,11 @@ <h1>Model Submission</h1>
             </p>
             <button type="submit" class="button submission">Submit Model</button>
         </form>
-        <p style="font-size: 12px"><a href="/profile">Return to profile page</a></p>
+        {% if "language" in request.path %}
+            <p style="font-size: 12px"><a href="/profile/language">Return to profile page</a></p>
+        {% else %}
+            <p style="font-size: 12px"><a href="/profile/vision">Return to profile page</a></p>
+        {% endif %}
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/upload.html:49]: use variable instead of if-else

File: benchmarks/urls.py [hunk 20]
```diff
@@ -1,22 +1,40 @@
 from django.urls import path
+import functools
 
 from .views import index, user, model, competition
 
 urlpatterns = [
     # index
-    path('', index, name='index'),
```
Comment (mschrimpf) [inline benchmarks/urls.py:7]: let's use a for loop over `['vision', 'language']` instead of a gigantic hand-written list

File: benchmarks/views/index.py [hunk 21]
```diff
@@ -31,15 +31,16 @@
 
 
 @cache_page(24 * 60 * 60)
-def view(request):
-    context = get_context()
+def view(request, domain):
```
Comment (mschrimpf) [inline benchmarks/views/index.py:34]:
```suggestion
def view(request, domain: str):
```

File: benchmarks/views/index.py [hunk 22]
```diff
@@ -31,15 +31,16 @@
...
+    context = get_context(domain=domain)
     return render(request, 'benchmarks/index.html', context)
 
 
-def get_context(user=None, benchmark_filter=None, model_filter=None, show_public=False):
-    benchmarks = _collect_benchmarks(user_page=True if user is not None else False,
+def get_context(user=None, domain="vision", benchmark_filter=None, model_filter=None, show_public=False):
```
Comment (mschrimpf) [inline benchmarks/views/index.py:39]:
```suggestion
def get_context(user=None, domain: str = "vision", benchmark_filter=None, model_filter=None, show_public=False):
```

File: benchmarks/views/index.py [hunk 23]
```diff
@@ -76,12 +77,12 @@ def get_context(user=None, benchmark_filter=None, model_filter=None, show_public
             "comparison_data": json.dumps(comparison_data)}
 
 
-def _collect_benchmarks(user_page=False, benchmark_filter=None):
+def _collect_benchmarks(domain, user_page=False, benchmark_filter=None):
```
Comment (mschrimpf) [inline benchmarks/views/index.py:80]:
```suggestion
def _collect_benchmarks(domain: str, user_page: bool = False, benchmark_filter=None):
```

File: benchmarks/views/index.py [hunk 24]
```diff
@@ -183,7 +184,7 @@ def _collect_submittable_benchmarks(benchmarks, user):
     return benchmark_selection
 
 
-def _collect_models(benchmarks, show_public, user=None, score_filter=None):
+def _collect_models(domain, benchmarks, show_public, user=None, score_filter=None):
```
Comment (mschrimpf) [inline benchmarks/views/index.py:187]:
```suggestion
def _collect_models(domain: str, benchmarks, show_public, user=None, score_filter=None):
```

File: benchmarks/views/user.py [hunk 25]
```diff
@@ -66,18 +72,36 @@ def post(self, request):
...
-            activation_link = f"{current_site}/activate/{uid}/{token}"
-            message = (f"Hello! Thanks for signing up with Brain-Score.\n\n"
-                       f"Please click or paste the following link to activate your account:\n{activation_link}\n\n"
-                       f"If you encounter any trouble, reach out to Martin (msch@mit.edu) or Mike (mferg@mit.edu)."
-                       f"Thanks,\n"
-                       f"The Brain-Score Team")
-            email = EmailMessage(mail_subject, message, to=[to_email])
-            email.send()
+
+            # if submitted via PR and now user is found, create a new user:
```
Comment (mschrimpf) [inline benchmarks/views/user.py:79]: could we make a unit test for this?

File: benchmarks/views/user.py [hunk 26]
```diff
@@ -153,7 +201,77 @@ def post(self, request):
         return render(request, 'benchmarks/success.html')
 
 
-def resubmit(request):
+def validate_zip(file):
```
Comment (mschrimpf) [inline benchmarks/views/user.py:204]: @kvfairchild: can we import a method from `core` to do this and avoid redundany?

**Review by mschrimpf — CHANGES_REQUESTED** [review]
There are a lot of places using `if language ... else vision` -- this is bad code, redundant, and doesn't scale. Use a generic variable instead. See comments for details.

File: benchmarks/urls.py [hunk 27]
```diff
@@ -1,22 +1,40 @@
...
+
+    # language changes
+    path('vision/', functools.partial(index, domain='vision'), name='index'),
+    path('language/', functools.partial(index, domain='language'), name='index'),
+    path('password/vision/',  user.Password.as_view(domain="vision"), name='password-vision'),
+    path('password/language/',  user.Password.as_view(domain="language"), name='password-language'),
+    path('password-change/vision/<str:uidb64>/<str:token>', user.ChangePassword.as_view(domain="vision"), name='change-password-vision'),
+    path('password-change/language/<str:uidb64>/<str:token>', user.ChangePassword.as_view(domain="language"), name='change-password-language'),
+    path('signup/vision', user.Signup.as_view(domain="vision"), name='vision-signup'),
+    path('signup/language', user.Signup.as_view(domain="language"), name='language-signup'),
```
Comment (mschrimpf) [inline benchmarks/urls.py:24]: why are there two separate signups? There's just one user table, let's only have `signup`.

File: benchmarks/templates/benchmarks/upload.html [hunk 28]
```diff
@@ -3,9 +3,24 @@
...
+            <h1>Submission System (Language)</h1>
+            Below you can upload a plugin. Plugins can be:
+                <ol>
+                    <li>A language model or revision</li>
+                    <li>A benchmark or revision</li>
+                    <li>A dataset or revision</li>
+                    <li>A metric or revision</li>
+                </ol>
+            See <a href="https://brain-score-core.readthedocs.io/en/latest/modules/glossary.html#term-Plug-in">
+        this template</a> for more information on plugin submissions.
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/upload.html:17]: this link does not describe the format.
@kvfairchild where should we point to?

File: benchmarks/templates/benchmarks/invalid_zip.html [hunk 29]
```diff
@@ -0,0 +1,18 @@
+{% extends 'benchmarks/base.html' %}
+
+{% block content %}
+<section id="success" class="container center">
+  <h2>Uh Oh! Your submission package was not in the correct format.</h2>
+  <p>
+    {{ error }}
+  </p>
+  <h3>The submission packaging rules can be found
+    <a href="https://brain-score.readthedocs.io/en/latest/">here</a></h3>
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/invalid_zip.html:10]: what's the correct link? @kvfairchild

File: benchmarks/templates/benchmarks/success.html [hunk 30]
```diff
@@ -2,11 +2,15 @@
...
 {% block content %}
 <section id="success" class="container center">
-	<p>Your submission was a success! Scores will be emailed to you whenever the tests complete.</p>
-  <p>Thank you for your participation!</p>
-  <br />
-  <p>To return to your profile page, click <a href="../profile">here</a>.</p>
-  <br />
-  <p>To submit another model, click <a href="../upload">here</a></p>
+  <h3>Your submission was a success! Scores will be emailed to you whenever the tests complete.</h3>
+  <h4>Thank you for your participation!</h4>
```
Comment (mschrimpf) [inline benchmarks/templates/benchmarks/success.html:6]: let's not use headings for everything, use messages: https://bulma.io/documentation/components/message/

File: benchmarks/urls.py (hunk 27 above)
Comment (mike-ferguson) [reply benchmarks/urls.py:24]: Will be addressed after new UI landing page (to provide domain - independent redirect)
